### ISSUE TITLE
Rule Request: Do not return NotImplemented from regular methods/functions

### ISSUE THREAD

**Issue by aardjon** [astral-sh/ruff#17011]
### Summary

[According to the docs](https://docs.python.org/3/library/constants.html#NotImplemented), `NotImplemented` is supposed to be mainly (or exclusively?) returned by some special dunder methods like `__eq__()`, `__lt__()` or `__imul__()`. That's why I would like to get a warning when returning it from any other context, because this may not be intended (and I consider it a code smell).

**Invalid Code:**

```python
def foo() -> bool:
    return NotImplemented
```

```python
class C:
    def __foo(self) -> str:
        return NotImplemented
```

**Valid Code:**
```python
class C:
    def __eq__(self, other: object) -> bool:
        return NotImplemented
```

I think in most cases one actually wants to raise instead, but I wouldn't fix it automatically because it changes behaviour.

As far as I know, no other linters can check this explicitly. Pylint produces follow-up issues if the returned value is not handled properly, Mypy can only warn about "Returning Any". But these warnings can be misleading.

**Comment by MichaReiser** [comment]
I think that makes sense to me, now that Python has abstract class (and method) support without needing `NotImplemented`. However, I don't think we should add this rule before we finish #1774 because I consider this a rule that intentionally restricts the allowed syntax.
